### ISSUE TITLE
Automatically hide or close OMEdit's Messages Browser (enhancement proposal)

### ISSUE THREAD

**Issue by ceraolo** [OpenModelica/OpenModelica#9889]
### Description
Today, when people work on has correctly running models, every simulation they make, even when it has finished successfuly,  they will see the precious work space filled for a large part  by the message browser window. 
If the simulation is successful the only important part of the Messages Browser content is "simulation finished successfully".
Personally, the first thing I do every time I have a successful simulation is to close the Message Browser, which after a few times becomes annoying. Keeping it afloat does not solve the issue.

I see two possible ways of improving this:

_Solution 1_ The Message browser is displayed only during the simulation, and when simulation finishes successfully it closes automatically. The sentence "simulation finished successfully" (or a shorter version of it, or even a green short rectangle) could conveniently placed in the status bar (in case of errors the Message Browser stays visible)
_Solution 2:_ the Message Browser is not displayed as default as compilation and execution proceeds. If this solution is adopted, the various phases of the process which terminates with simulations, shoud be monitored by simple info in the status bar, such as "analysing" and "compiling" before simulation starts, a short progress bar during simulation, "simulation successful" upon successful termination. In case of errors the Message Browser opens showing its content with focus on the error message.

I personally prefer solution 2, but solution 1 is also ok for me.

Obviously, for both solutions in case the users want to have a look at messages, even in case of successful simulations, they can access the message browser through the already existing View|Windows menu item.

I took inspiration for these proposals (solution 1 and 2) from the behaviour of Qt Creator, which, although being a professional tool, leaves during compilation the workspace free for user's activity, while still allowing to gain access, in case of need, to all the info on what has happened during compilation.

### Version and OS
This is not a bug report, but an enhancement suggestion. 
  I decided to post this ticket now, after having seen in 1.21.0-beta that OM is good enough to be perceived by the final users as a mature product, that hides unnecessary details in case of succesful simulations, letting them to concentrate on their work

**Comment by adeas31** [comment]
We support multiple parallel simulations and both solutions doesn't fit for that.

One solution is to just show a horizontal bar with buttons that shows the names of the tabs that are in `Messages Browser` and when user clicks on the button we open the clicked tab in Messages Browser. This is one idea, I am not sure even if its possible to implement.

**Comment by ceraolo** [comment]
I strongly believe the ideas proposed in this ticket would greatly improve the experience of the large majority of users, who, I think, do not use parallel simulations or rarely use them.

Therefore I propose to evaluate two options (I do not have preference over them) to keep this ticket's proposals alive:
1. adopt one of this ticket's proposals the first simulation. I mean the unique simulation in case there exists just one, the first one of multiple ones in the other case. The second, third, etc. of parallel simulations would behave as they do already now
2. use a flag to choose which way the messages are managed: with (as it is now) or without (as it is proposed in this ticket) a message browser window.

**Comment by casella** [comment]
@ceraolo we are planning to restructure the setup interface in the next few months. The idea is to separate "basic" from "advanced" stuff. We can also keep this requirement of yours into consideration.

**Comment by ceraolo** [comment]
This is very near to my solution 1. 

If my solution 1 is adopted, I would insist a little in asking to let "simulation finished successfully" appear in the status bar, instead of having to keep the message browser open a couple of seconds. Timed interaction with users often has annoying side-effects. Using status bar for very concise messages is a standard solution for many programs, and should be very easy to implement.

Just for completeness. My solution 2 was more innovative, and did not require any loss of information for people having will to read all messages. After all it was taken from Qt Creator, which is a professional tool for software authors. It simply a matter of allowing interested users to have easy access to message browser output possibly taking inspiration from Qt creator

**Comment by casella** [comment]
If the message browser is disabled, and if the simulation ends successfully, I am fine at having just the progress bar and a simple message showing up in the status bar. I guess this is what you need for unproblematic models, when you re-simulate often changing parameters every time, and you don't want to be bothered by that.

But if one wants to always see the messages window, I would keep it open. I am used to this setting, I find it perfectly fine and I really see no reason to change it.

**Comment by ceraolo** [comment]
> If the message browser is disabled, and if the simulation ends successfully, I am fine at having just the progress bar and a simple message showing up in the status bar. I guess this is what you need for unproblematic models, when you re-simulate often changing parameters every time, and you don't want to be bothered by that.
> 
> But if one wants to always see the messages window, I would keep it open. I am used to this setting, I find it perfectly fine and I really see no reason to change it.

I totally agree. I think both usage models should be allowed.

**Comment by casella** [comment]
@adeas31 we can discuss this in the next devmeeting.

**Comment by casella** [comment]
@ceraolo we had some discussion and we think we found a good solution.

OMEdit will internally keep a visualization status of the Messages Browser (MB). This status is determined when you start a simulation: if the MB happens to be open, then the status is set to on, otherwise it is set to off. The message browser will then be always displayed as the simulation(s) are running. When it/they are all terminated successfully, the MB will remain open if the status is on, otherwise it will be closed automatically. If at least one of the simulations fails, then it will remain open, so the user can inspect the reasons of failure.

The message "Simulation of xxx finished successfully" will be put in the status bar when a simulation ends, if there are multiple ones running in parallel, the second successfully finishing simulation will overwrite the first and so on. Ditto for simulation of xxx failed. We can't put it on the bottom left corner, which is already used to display some service messages (see e.g. the description of menu items when you make a menu selection). We can use the area to the left of line/column for that.

@ceraolo I understand this perfectly fits your needs, as well as mine. Please confirm.

**Comment by casella** [comment]
@adeas31 a couple more things while we're at this. Consider these two MWE:
```
model M1
Real x(start = 1, fixed = true);
equation
  der(x) = -1/(x-0.5);
end M1;

model M2
Real x(start = 1, fixed = true);
equation
  der(x) = 1/(x-0.5);
end M2;
```
If you simulate the first, the simulation aborts at time 0.125 because of a singularity. In this case, the output tab is correctly set to display the first lines of the simulation log, because in case of errors you need to look at them from the beginning.

If you simulate the second, instead, the simulation runs successfully until the end. In this case, if the Output tab window is not large enough to fit the whole log, I think it would be best to pre-set the slider at the bottom position, so that you clearly see the "The simulation finished successfully" message, and you can expand the statistics right away. Currently, you have to scroll down to see what it the end result, which is a bit annoying.

Last, but not least, it would be nice if the message that appears on the top left of the MB and will appear in the status bar for M1 would be "Simulation of M1 failed", rather than "Simulation of M1 is finished", based on the return value of the simulation executable. Is this possible?

**Comment by adeas31** [comment]
@ceraolo @casella I tried implementing the solution we discussed but the result is very messy (auto opening and closing of Messages Browser) and misleading to the user in case of parallel simulations.

I come up with a new solution. I created a small tab widget that looks very similar to Messages Browser. It will only show the tabs and simulation information.

![image](https://github.com/OpenModelica/OpenModelica/assets/4007231/ee9b2221-a0d8-4873-a85d-386ed8299a1a)

The tab widget is only shown when Messages Browser is hidden. If Messages Browser is not visible when starting the simulation then the information will be shown in this tab widget but you can always click on it to show the full Messages Browser.

This solution allows us to keep the parallel simulations, also gives more space for plotting etc. And if you always have Messages Browser open then you will never notice this tab widget.

The Messages Browser will automatically open when there is a notification, warning or error message. This is same as before.

The only annoying thing is that in normal view when Messages Browser is not visible. You will see,

![image](https://github.com/OpenModelica/OpenModelica/assets/4007231/9074c3ba-21c4-423c-aca5-06ca1dd8ee54)

Note the message "Click to open Messages Browser" at the bottom. This will be shown always. The tab widget has a height of 2 rows to accommodate the simulation message and progress. If I don't show the message then we will see the empty space. So I thought its better to show a message.

Let me know what you think about this. You can try it out in the tomorrow's nightly.

If we want to implement the solution we discussed earlier then I suggest we disable parallel simulations. I am not sure how much that feature is used by the users but I think it is good to have in case you run long simulations.

**Comment by adeas31** [comment]
> If you simulate the first, the simulation aborts at time 0.125 because of a singularity. In this case, the output tab is correctly set to display the first lines of the simulation log, because in case of errors you need to look at them from the beginning.

> If you simulate the second, instead, the simulation runs successfully until the end. In this case, if the Output tab window is not large enough to fit the whole log, I think it would be best to pre-set the slider at the bottom position, so that you clearly see the "The simulation finished successfully" message, and you can expand the statistics right away. Currently, you have to scroll down to see what it the end result, which is a bit annoying.

This was a bug. We don't scroll he slider when new output is added. I have fixed it now. But if you want to scroll to top in case of error then of course we can do that.

> Last, but not least, it would be nice if the message that appears on the top left of the MB and will appear in the status bar for M1 would be "Simulation of M1 failed", rather than "Simulation of M1 is finished", based on the return value of the simulation executable. Is this possible?

Should be fixed with #11627.

**Comment by ceraolo** [comment]
> 
> Let me know what you think about this. You can try it out in the tomorrow's nightly.
> 


I'll check by the end of the week. 
Thanks.

**Comment by casella** [comment]
@adeas31 looks very good to me from what you are showing here, I'll check that ASAP. Parallel simulations can be very useful if you are exploring the parameter space of a complex model, I would not disable that if we already support it.

Regarding the scrolling, you are right it's a bug, in fact I think I remember it once scrolled until the end. That should be the preferred behaviour. If you can scroll it back to the beginning in case of errors, I think it would be good for beginners, because you should always check error message logs from the top (where presumably the root cause lies), not from the bottom.

Thanks!

**Comment by ceraolo** [comment]
I'm not sure I understand your modification well.
The purpose of this ticket was to avoid using desktop area with Message browser when simulations run smooth. Ideally when they run smoothly I could just have a textual row in the status bar saying:
- > compiling
- > running (or simulating)
- > simulation finished successfully.

When something goes wrong, or upon user's request, the message browser should be displayed with full info.

From today's nightly (dev 151, tested on windows) It seems that I still have the message browser that opens and stays open when the simulation is fhinshed.
Do I miss something?

**Comment by adeas31** [comment]
If you close Messages Browser and start the simulation then it should not be shown. Instead you will see an alternate tab widget which is a small version of Messages Browser (which takes only 2 rows).

**Comment by ceraolo** [comment]
at least in 1.23-dev151 in Linux, even though I close it, it opens again in the next simulation.

**Comment by adeas31** [comment]
The Linux nightly build are broken at the moment. Makesure you have this commit https://github.com/OpenModelica/OpenModelica/commit/ddbb85bbe309aa2e77a56a7ddd4f6a77f4561efc

**Comment by casella** [comment]
@ceraolo when checking bleeding edge updates, please always do `omc --version`, then check the hash code against the [list of commits](https://github.com/OpenModelica/OpenModelica/commits/master),  making sure that the commit that fixed the issue is included.

**Comment by ceraolo** [comment]
Sorry, this is too much for me.
What I can do is to wait a few more days before downloading nightly builds in my Linux VM.

**Comment by casella** [comment]
Seriously, it's no big deal. omc --version gives you something like `OpenModelica v1.23.0-dev-151-g8fef40a3f9`. The code after the -g (in this case 8fef40a3f9) is the hash code.

The commit list is something like this:
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/12be450c-6d16-4dc6-b68d-92b415820d58)

On the right you can see the hash codes. @adeas31's commit #11627 is clearly visible. In my case `8fef40a3f9` doesn't show up there, or in any later commit, so clearly my nightly build was compiled before that commit was made. 

That's it 😃 

For windows nightlies, the has code is actually provided in the installer filename, so it's even easier to check if it includes the commit you want to test or not.

**Comment by ceraolo** [comment]
Even simlpler, we see the hash code also in the menu OMEdit Help|About OMEdit.
What' still confusing me, also in your example, is that the hash code is 8fef40a3f9, much longer than those in the list. How did you select 8fef40a?

**Comment by casella** [comment]
The GIT has code is 40 characters long, which makes the likelyhood of two different pieces of code having the same hash code really, really low. 

In real life, you don't really need all of those characters to guarantee uniqueness - if you are checking a handful of commits, the first 5, 6, or 7 digits are more than enough to uniquely identify it. How many exactly, it doesn't really matter.

**Comment by ceraolo** [comment]
@adeas, I 90%. like your solution.

I see that Messages Browser (BTW, why not to call it "Message Browser"?) now opens whenever you simulate a new model, and stays open until you close it.
If you close it, it stays closed when re-simulating (that is very good)

Ideally, IMO, 
**after you've closed it,** 
_it should stay closed up to when the user clicks on the text in the simulation tab "Simulation of [....] finished" or simulation is unsuccessful._ 
Or, if you don't like this, 
_when simulating a new model it can open, but should close automatically when simulation is successful._

this behaviour should be kept until the user voluntarily opens Messages Browser again 

@adeas, @casella, what do you think? I suppose that this further change should be very easy to implement.

**Comment by adeas31** [comment]
You tagged someone else :).

> I see that Messages Browser (BTW, why not to call it "Message Browser"?) now opens whenever you simulate a new model, and stays open until you close it.

All the browsers are called like this in OMEdit e.g., `Libraries Browser`, `Variables Browser`. Anyways, the MB should not open when you simulate a model. If it is already open then its kept. If its not open then you will get the updates in the new small tab widget.

**Comment by adeas31** [comment]
That's weird. It doesn't happen for me. Check the following video.

https://github.com/OpenModelica/OpenModelica/assets/4007231/c616a691-5bb1-40ed-b6cd-26c01526e80b

Maybe you have some translation errors in your model. If that is the case then the MB will open up to show those errors.

**Comment by adeas31** [comment]
Ported this to 1.22.1 #11655

**Comment by ceraolo** [comment]
Yes, I had the following warning:
![image](https://github.com/OpenModelica/OpenModelica/assets/7373030/d9e1e374-6154-4fbd-90bb-99af1a522835)
---
Adding "fixed=true" allowed Browser staying closed.

So, the current behaviour is that Browser opens when there is at least a warning.

If this is what you intended, this is ok also for me. 

It forces us to read not only errors but also warnings, and lets us free from always getting an open Messages Browser window when we re-simulate.

**Comment by adeas31** [comment]
Yeah so I guessed it right 😉. The browser will open up if there is any translation notification/warning/error.

**Comment by casella** [comment]
Yeah, this may be overly cautious, but I guess one should always consider warnings.

**Comment by casella** [comment]
Regarding the language issue, what I learned in school is that nouns used as adjectives should not take the plural, so it should indeed be "Message Browser" and "Library Browser", but I'm not a native speaker either - maybe using the plural is also accepted nowadays. I will ask a (very) British friend of mine and report.

**Comment by ceraolo** [comment]
Nowadays plural is accepted in some cases, I would say.
For instance "Ryanair, the low fares airline" (not fare).
But English books and teachers still say that it is mandatory to say "tooth brush" (not teeth brush), "shoe shop" (not shoes shop) etc. 

So as you say, it is a topic in evolution.  Let us see what the very British guy says.

**Comment by casella** [comment]
> So as you say, it is a confusing topic, maibe in evolution. Let us see what the very British guy says.

He replied me in Italian (he lives in Venice and has a PhD in baroque opera, currently trying to rebuild [Teatro S. Cassiano](https://www.teatrosancassiano.it/):
> indiscutibilmente "message browser"
which translates to 
> unquestionably "message browser"

I'll open a ticket about that.

**Comment by ceraolo** [comment]
I did not dare being too explicit, but frankly my ears like this much more than Messages Browser.

**Comment by ceraolo** [comment]
After some weeks of usage of OM with this ticket closed, I see that there are several cases for which warnings are unavoidable, or acceptable, or would require a lot of extra work to avoid.

Therefore, I would appreciate if MessageBoxes are not automatically reopened when re-simulating, at least for people like me.

**I propose to add in the mesage box a checkbox of the type
"do not automatically reopen MessageBox upon successful re-simulation"**

Checking this checkbox would set some internal flag that can be possibly reversed through Tools|Options menu.
In this way, the modification will not alter the current behaviour for those that like it, but will allow those that know that their models imply acceptable warnings to concentrate on their work.

@casella @adeas31 Do you think this is possible?

An interesting case-study for this message's request is when solving Power Flow probems with the latest PowerGrids. One wold need to change PV voltages or other parameters several times, and would like to see the effect of their changes on the diagram. This is very fast for the simulation part, but requires time to close MessageBoxes and close and reopen diagram (for which I've already opened #11811). 
I would say most of the time is spent opening and closing windows. Addressing this comment and #11811 would ease users's life a lot, for this case-study.

Thank you.

**Comment by ceraolo** [comment]
I think this ticket should be reopened.
Consider this:
[RL.mo.txt](https://github.com/OpenModelica/OpenModelica/files/13981406/RL.mo.txt)
This runs without any warnings. However, OMEdit keeps opening MessageBrowser over and over again at any simulation.
Tested with  v1.22.0-dev-341

**Comment by casella** [comment]
The proposal seems reasonable to me. I'm not sure why the RL example reopens the Message Browser, @adeas31 can you please check?

**Comment by ceraolo** [comment]
> The proposal seems reasonable to me. 

**Maybe wven better (and easier to implement) would be as follows** 

add in the Message Browser a checkbox of the type
"do not automatically reopen Message Browser upon successful simulation"

Checking this checkbox would set some internal flag that can be possibly reversed through Tools|Options menu.

In case of warnings, some visual clickable hint on the MessageBrowser bar at the lower part of the screen could be set (e.g. a yellow square). If the user want to see the warnings they have juust to click on that clickable hint, or other control around there to open MessageBrowser

**Comment by adeas31** [comment]
> The proposal seems reasonable to me. I'm not sure why the RL example reopens the Message Browser, @adeas31 can you please check?

The RL example works as expected for me. See the screencast below. I first simulated and re-simulated the model with Message Browser closed and then simulated and re-simulated with Message Browser closed. The simulation doesn't cause the Message Browser to open/close. As I said earlier, if you have any translation message (error, warning or notification) then the Message Browser will open.

https://github.com/OpenModelica/OpenModelica/assets/4007231/5e2f6096-8cce-4d76-8340-05f034206225

Regarding @ceraolo proposal, I believe we should not add the checkbox to Message Browser. The checkbox should be only available in `Tools|Options|Messages` with text `Do not automatically enlarge Message Browser when a new message is available`. The checkbox should be unchecked by default.

**Comment by ceraolo** [comment]
thank you @adeas31 for considering this ticket again.

1) regarding RL: I could not reproduce what I posted. So it is my mistake, as you say, this file RL.mo works as expected.

2) regarding the checkbox: ok for putting it in Tools|Options|Messages as you say.
Only, I think that  the collapsed message browser should give some hint on whether warnings or errors are present. Maybe a green/yellow/red small area for  "no warnings/errors", "only warnings", "errors".
So,  users will know what content the enlarged Message Browser will contain.

**Comment by adeas31** [comment]
@ceraolo PR #11970 does what we discussed above. The collapsed message browser gives hint with the blinking text.

Try it with tomorrow's nightly.

**Comment by ceraolo** [comment]
It seems to me that after an error, MessageBrowser continues to open, even if no error occurs.

Examples in the following video:
[Screencast from 09-02-2024 15:49:08.webm](https://github.com/OpenModelica/OpenModelica/assets/7373030/75c0dda0-7186-4855-bb8b-aa3f303fe532)

Here the error is that I put a negative resistance value in a diode. When I restore the correct resistance, MessageBrowser continues to pop-up even if I continue to close it and even if no error is anymore present in the model.

**Comment by adeas31** [comment]
Have you enabled the setting in `Do not automatically enlarge Message Browser when a new message is available` in `Tools|Options|Messages`?

**Comment by ceraolo** [comment]
I wanted first to check the behaviour without that new flag.
So, the behaviour _without the new flag_ is that MessageBrowser is opened even without errors or warnings in case of "true" simulation (not re-simulation). It does not open only in case of resimulations without errors/warnings. Now I got it.

Now I selected that new checkbox.
I tried the model [OneSQWRe.mo.txt](https://github.com/OpenModelica/OpenModelica/files/14224333/OneSQWRe.mo.txt), which has errors.
When I run it with MessageBrowser closed, I see "All" and "Warnings" labels blinking, while the simulation failed with an error (see below). This is not optimal: when the user does not see "Errors" blinking, they expect to have successful simulations I would say.

Moreover, the progress bar indicates 19% which is misleading, since the simulation did not even start.

![Screenshot from 2024-02-09 18-04-05](https://github.com/OpenModelica/OpenModelica/assets/7373030/fa59b233-c84b-4812-990e-ed01cb1edb99)

**Comment by ceraolo** [comment]
Thank you, @adeas!
I'll check tomorrow.
I have another (very minor) change to ask on this, but I don't want to be too annoying. I'll wait till tomorrow.

Thanks again on your support on this ticket. I realise that not many people feel the need of this, but my experience of several hours per day on several days of the year, suggests that this work of yours will have good effects on all the serious users of OM.

**Comment by ceraolo** [comment]
MMM. I suppose this was closed a bit too early... We'll see tomorrow.

**Comment by casella** [comment]
And?

**Comment by adeas31** [comment]
> I wanted first to check the behaviour without that new flag. So, the behaviour _without the new flag_ is that MessageBrowser is opened even without errors or warnings in case of "true" simulation (not re-simulation). It does not open only in case of resimulations without errors/warnings. Now I got it.

Yes. I believe we should enlarge the Message Browser when a new simulation is started and the flag is not set.

> 
> Now I selected that new checkbox. I tried the model [OneSQWRe.mo.txt](https://github.com/OpenModelica/OpenModelica/files/14224333/OneSQWRe.mo.txt), which has errors. When I run it with MessageBrowser closed, I see "All" and "Warnings" labels blinking, while the simulation failed with an error (see below). This is not optimal: when the user does not see "Errors" blinking, they expect to have successful simulations I would say.

The tabs `All`, `Notifications`, `Warnings` and `Errors` are for general messages from the application and the translation errors from the compiler. The error messages related to the simulation will appear in its own tab instead of `Errors` so there is no point in blinking that tab since there is nothing new to show. In this case you should see the text `Simulation of OneSQWRe failed.` next to the progress bar. I will investigate why you are not seeing that message.

![image](https://github.com/OpenModelica/OpenModelica/assets/4007231/6f0ad539-9d27-48bd-b702-a4631dbe92fc)

> 
> Moreover, the progress bar indicates 19% which is misleading, since the simulation did not even start.
> 

OMEdit receives the simulation progress from the simulation executable. If you believe this 19% is wrong then I suggest creating a separate issue for it.

**Comment by ceraolo** [comment]
> In this case you should see the text `Simulation of OneSQWRe failed.` next to the progress bar. I will investigate why you are not seeing that message.

Thank you for this
 
> OMEdit receives the simulation progress from the simulation executable. If you believe this 19% is wrong then I suggest creating a separate issue for it.

My fault, progress bar is ok.

Just a simple addition. 
In "computerese" blinking is used for requiring immediate attention from the user.
The idea of keeping MessageBrowser collapsed is that when the user knows that ther model have warnings (for instance when initial conditions are not all set) they will not open Messager Browser. Having blinking text is uncomfortable, becasue, as I said, it means that immediate user action is required.

Therefore I would kindly ask to use a different way of indicating that warnings or errors are present. Maybe they can be changed into italic, as OMEdit usually does when some model file are changed? Maybe italic with the addition of an asterisk? Or even a change in the tesxt color into yellow or red?

**Comment by adeas31** [comment]
> Therefore I would kindly ask to use a different way of indicating that warnings or errors are present. Maybe they can be changed into italic, as OMEdit usually does when some model file are changed? Maybe italic with the addition of an asterisk? Or even a change in the tesxt color into yellow or red?

It is not possible to change the font of the text. Would it be enough just with a asterisk?

**Comment by ceraolo** [comment]
An asterisk is ok for me.Il 16 Feb 2024 15:58, Adeel Asghar ***@***.***> ha scritto:

Therefore I would kindly ask to use a different way of indicating that warnings or errors are present. Maybe they can be changed into italic, as OMEdit usually does when some model file are changed? Maybe italic with the addition of an asterisk? Or even a change in the tesxt color into yellow or red?

It is not possible to change the font of the text. Would it be enough just with a asterisk?

**Comment by casella** [comment]
@adeas31 I'm re-opening this until the asterisk is fixed 😃
